src/tagged-type.ts

```diff
@@ -0,0 +1,141 @@
+export type Module = string;
+export type DocumentReference = string;
+
+/**
+ * TaggedType provides the encoding/decoding of the Fauna Tagged Type formatting
+ */
+export class TaggedTypeFormat {
+  /**
+   * Encode the Object to the Tagged Type format for Fauna
+   *
+   * @param obj - Object that will be encoded
+   * @returns Map of result
+   */
+  static encode(obj: any): any {
+    return new TaggedTypeEncoded(obj).result;
+  }
+
+  /**
+   * Decode the JSON string result from Fauna to remove Tagged Type formatting.
+   *
+   * @param input - JSON string result from Fauna
+   * @returns object of result of FQL query
+   */
+  static decode(input: string): any {
+    return JSON.parse(input, (_, value: any) => {
+      if (value["@mod"]) {
+        return value["@mod"] as Module;
+      } else if (value["@doc"]) {
+        return value["@doc"] as DocumentReference;
+      } else if (value["@int"]) {
+        return Number(value["@int"]);
+      } else if (value["@long"]) {
+        return BigInt(value["@long"]);
+      } else if (value["@double"]) {
+        return Number(value["@double"]);
+      } else if (value["@date"]) {
+        return new Date(value["@date"] + "T00:00:00.000Z");
+      } else if (value["@time"]) {
+        return new Date(value["@time"]);
+      } else if (value["@object"]) {
+        return value["@object"];
+      }
+
+      return value;
+    });
+  }
+}
+
+type TaggedDate = { "@date": string };
+type TaggedDouble = { "@double": number };
+type TaggedInt = { "@int": number };
+type TaggedLong = { "@long": string };
+type TaggedObject = { "@object": Record<string, any> };
+type TaggedTime = { "@time": string };
+
+class TaggedTypeEncoded {
+  readonly result: any;
+
+  readonly #encodeMap = {
+    bigint: (value: bigint): TaggedLong => {
+      if (value >= -(2 ^ 63) + 1 && value <= 2 ** 63 - 1) {
+        return {
+          "@long": value.toString(),
+        };
+      }
+      throw new TypeError("Precision loss when converting int to Fauna type");
+    },
+    number: (value: number): TaggedDouble | TaggedInt | TaggedLong => {
+      if (`${value}`.includes(".")) {
+        return { "@double": value };
+      } else {
+        if (value >= -(2 ^ 31) + 1 && value <= 2 ** 31 - 1) {
```

Comment:
`^` is the XOR operator. This looks like a bug.

Also, 2's complement is [- 2^(n-1), 2^(n-1) - 1]. Our range is incorrect.

We are also missing tests that would have caught these issues.

Let's add some basic tests to test number ranges and mappings.